src/engine/GroupBy.cpp

```diff
@@ -763,25 +763,35 @@ GroupBy::findAggregates(sparqlExpression::SparqlExpression* expr) {
 
 // _____________________________________________________________________________
 template <class T>
-bool GroupBy::hasType(sparqlExpression::SparqlExpression* expr) {
-  return dynamic_cast<T*>(expr) != nullptr;
+std::optional<T*> GroupBy::hasType(sparqlExpression::SparqlExpression* expr) {
+  auto value = dynamic_cast<T*>(expr);
+  if (value == nullptr)
+    return std::nullopt;
+  else
+    return value;
 }
 
 // _____________________________________________________________________________
 bool GroupBy::isSupportedAggregate(sparqlExpression::SparqlExpression* expr) {
   using namespace sparqlExpression;
 
-  return !(hasType<SumExpression>(expr) || hasType<MinExpression>(expr) ||
-           hasType<MaxExpression>(expr) || hasType<CountExpression>(expr) ||
-           hasType<GroupConcatExpression>(expr));
+  return !(hasType<SumExpression>(expr).has_value() ||
+           hasType<MinExpression>(expr).has_value() ||
+           hasType<MaxExpression>(expr).has_value() ||
+           hasType<CountExpression>(expr).has_value() ||
+           hasType<GroupConcatExpression>(expr).has_value());
```

Comment:
I think you can leave it as it was (the optional has an implicit conversion to bool, and your function already has the "booleansounding" name.

Reply:
And at a second glance:
You can (already now)
refactor this function to be something like:
1. The function is called `isUnsupportedAggregate` and the return type is inverse.
2. The function checks whether it is an aggregate that is either not AVG or not `distinct`.
3. For this you need an `isAggregate()` function in the sparqlExpressions. Best implement it, and then make the `containsAggregate` a `final` function that is implemented in terms of `isAggregate()`. That makes the code much simpler.